sql/sem: add check for interval for asof.DatumToHLC()

ZhouXing19 opened the pull request:
Currently, if the given interval for `AS OF SYSTEM TIME interval` is a small postive duration, the query can incorrectly pass. It's because when we call `clock.Now()`, it has passed the threashold ('statement timestamp + duration').

Now we add a check for the duration value. It has to be negative, with the absolute value greater than a nanosecond.

fixes #91021
link epic CRDB-17785

Release note (bug fix): add restriction for duration value for AS OF SYSTEM TIME statement.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/93146)
<!-- Reviewable:end -->

rafiss reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93146)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ZhouXing19)

___
*[`pkg/sql/sem/asof/as_of.go` line 253 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93146#-NIhNLUM-gIBlzhcD520:-NIhNLUN7AKkNinEdrgv:b5zv1lx) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd07b13eff1f9ab9ebf287865697d10018d48fcd/pkg/sql/sem/asof/as_of.go#L253)):*
> ```Go
> 		if iv, err := tree.ParseDInterval(evalCtx.GetIntervalStyle(), s); err == nil {
> 			if (iv.Duration == duration.Duration{}) {
> 				convErr = errors.Errorf("interval value %v too small, absolute value must be >= %v", d, time.Microsecond)
> ```

is `time.Microsecond` correct in this earlier message?

actually, we can just remove this `if` case if we change the below ones to `iv.Duration.Compare(duration.Duration{}) >= 0` and `iv.Duration.Compare(duration.Duration{}) <= 0`
___
*[`pkg/sql/sem/asof/as_of.go` line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93146#-NIhNTGM40h_fhg0ZeyF:-NIhNTGM40h_fhg0ZeyG:b-4sknki) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd07b13eff1f9ab9ebf287865697d10018d48fcd/pkg/sql/sem/asof/as_of.go#L255)):*
> ```Go
> 				convErr = errors.Errorf("interval value %v too small, absolute value must be >= %v", d, time.Microsecond)
> 			} else if (usage == AsOf && iv.Duration.Compare(duration.Duration{}) > 0 && !syn) {
> 				convErr = errors.Wrapf(
> ```

i think this should just be:
```
convErr = errors.Newf("interval value %v too large, AS OF interval must be <= %v", d, -time.Nanosecond)
```
___
*[`pkg/sql/sem/asof/as_of.go` line 261 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93146#-NIhPQmf56fl4MdhVzog:-NIhPQmf56fl4MdhVzoh:b-7d98at) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd07b13eff1f9ab9ebf287865697d10018d48fcd/pkg/sql/sem/asof/as_of.go#L261)):*
> ```Go
> 				)
> 			} else if (usage == Split && iv.Duration.Compare(duration.Duration{}) < 0 && !syn) {
> 				convErr = errors.Wrapf(
> ```

i think this should just be:
```
convErr = errors.Newf("interval value %v too small, SPLIT AT interval must be >= %v", d, time.Nanosecond)
```
___
*[`pkg/sql/sem/asof/as_of.go` line 290 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93146#-NIhPfinCpY7_RYz0ugH:-NIhPfinCpY7_RYz0ugI:bnnslav) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd07b13eff1f9ab9ebf287865697d10018d48fcd/pkg/sql/sem/asof/as_of.go#L290)):*
> ```Go
> 				errors.Newf("unsupported interval duration: %v", d),
> 				"expiration interval must be non-negative",
> 			)
> ```

same comment on error messages (and using `<= 0` and `>= 0` in the `if`s


<!-- Sent from Reviewable.io -->

ZhouXing19 committed:
sql/sem: add check for interval for asof.DatumToHLC()

  fixes #91021

  Currently, if the given interval for `AS OF SYSTEM TIME interval` is a small
  postive duration, the query will incorrectly pass. It's because when we call
  `clock.Now()`, it has passed the threashold ('statement timestamp + duration').

  Now we add a check for the duration value. It has to be negative, with the
  absolute value >= a microsecond.

  Similarly, when the logic is used under the SPLIT stmt, the interval's value has
  to be positive (for the experation duration), with an absolute value >= a
  microsecond.

  link epic CRDB-17785

  Release note (bug fix): add restriction for duration value for AS OF SYSTEM TIME
  and SPLIT statement.

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93146)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/sql/sem/asof/as_of.go` line 253 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93146#-NIhNLUM-gIBlzhcD520:-NIhubkB2T1Ft2NETqf9:brhszkg) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd07b13eff1f9ab9ebf287865697d10018d48fcd/pkg/sql/sem/asof/as_of.go#L253)):*
> is time.Microsecond correct in this earlier message?

Ah yes, it was correct, I made a typo. Changed them back.

>  change the below ones to iv.Duration.Compare(duration.Duration{}) >= 0 and iv.Duration.Compare(duration.Duration{}) <= 0

I separated the `==` case because we always disallow it, but we actually allow `>` for `AS OF` and `<` for `SPLIT` if the interval is synthetic ,see [here](https://github.com/cockroachdb/cockroach/blob/593bcb23dc422bcb3cc9ea40a36d75c432042756/pkg/sql/as_of_test.go#L146). 
___
*[`pkg/sql/sem/asof/as_of.go` line 255 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93146#-NIhNTGM40h_fhg0ZeyF:-NIhv2ip6BqnpcFQLBYk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd07b13eff1f9ab9ebf287865697d10018d48fcd/pkg/sql/sem/asof/as_of.go#L255)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think this should just be:
```
convErr = errors.Newf("interval value %v too large, AS OF interval must be <= %v", d, -time.Nanosecond)
```
</blockquote></details>

Done.
___
*[`pkg/sql/sem/asof/as_of.go` line 261 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93146#-NIhPQmf56fl4MdhVzog:-NIhxaEK2FrwkM_V-ogD:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd07b13eff1f9ab9ebf287865697d10018d48fcd/pkg/sql/sem/asof/as_of.go#L261)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

i think this should just be:
```
convErr = errors.Newf("interval value %v too small, SPLIT AT interval must be >= %v", d, time.Nanosecond)
```
</blockquote></details>

Done.
___
*[`pkg/sql/sem/asof/as_of.go` line 290 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/93146#-NIhPfinCpY7_RYz0ugH:-NIhxxGT4uoNQibo3EIG:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd07b13eff1f9ab9ebf287865697d10018d48fcd/pkg/sql/sem/asof/as_of.go#L290)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

same comment on error messages (and using `<= 0` and `>= 0` in the `if`s
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

rafiss reviewed (approved):
thanks for the fix!

> Reviewed 1 of 4 files at r1, 2 of 3 files at r3, 1 of 1 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93146#-:-NInyP3xAuUsMVOSXjMl:bp9cthg)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ZhouXing19)

___
*[`-- commits` line 11 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/93146#-NInzcFKEU8d_v4gJ9Sk):*
nit: commit message still uses nanosecond instead of microsecond
___
*[`pkg/sql/sem/asof/as_of.go` line 256 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/93146#-NInzvqHCAOPIen9vUHa:-NInzvqHCAOPIen9vUHb:b-b5putm) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb12bca858784bc520c59478a93e850baaa597c7/pkg/sql/sem/asof/as_of.go#L256)):*
> ```Go
> 			} else if (usage == AsOf && iv.Duration.Compare(duration.Duration{}) > 0 && !syn) {
> 				convErr = errors.Errorf("interval value %v too large, AS OF interval must be <= -%v", d, time.Microsecond)
> 			} else if (usage == Split && iv.Duration.Compare(duration.Duration{}) < 0 && !syn) {
> ```

are you sure the `syn` check is needed on this one? (i'm not too familiar with it)


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/93146)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`-- commits` line 11 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/93146#-NInzcFKEU8d_v4gJ9Sk):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

nit: commit message still uses nanosecond instead of microsecond
</blockquote></details>

Done.
___
*[`pkg/sql/sem/asof/as_of.go` line 256 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/93146#-NInzvqHCAOPIen9vUHa:-NIrq3dd7G5t1imXcVNt:b-uassow) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb12bca858784bc520c59478a93e850baaa597c7/pkg/sql/sem/asof/as_of.go#L256)):*
<details><summary><i>Previously, rafiss (Rafi Shamim) wrote…</i></summary><blockquote>

are you sure the `syn` check is needed on this one? (i'm not too familiar with it)
</blockquote></details>

Good point, I'm not familiar either, and there wasn't check / test for synthetic timestamp previously for SPLIT AT. I removed it for now and added a comment.


<!-- Sent from Reviewable.io -->

ZhouXing19 commented:
TFTR!
bors r=rafiss
